qiskit_optimization/algorithms/grover_optimizer.py

```diff
@@ -246,15 +284,19 @@ def solve(self, problem: QuadraticProgram) -> OptimizationResult:
                     threshold = optimum_value
 
                     # trace out work qubits and store samples
-                    if self._quantum_instance.is_statevector:
-                        indices = list(range(n_key, len(outcome)))
-                        rho = partial_trace(self._circuit_results, indices)
-                        self._circuit_results = cast(Dict, np.diag(rho.data) ** 0.5)
-                    else:
+                    if self._sampler:
```

Comment:
This should be an explicit test against None I think too.

Reply:
Do you mean the following?
```suggestion
                    if self._sampler is not None:
```